### PR TITLE
Socket api fixes

### PR DESCRIPTION
**Pull request by folkertdev**

### CODE REVIEW

**Comment by davidv1992** [comment]
Outside of the comment formatting, I am not a big fan of these changes. I think most of what is done here should really be part of a bigger push to get the interface portion of the library in good shape, rather than half-fixes here.

**Comment by folkertdev** [comment]
I'm confused by that. Nothing interesting is really happening here, it's just clarifying code in some places. Do not let the perfect be the enemy of the good

**Comment by davidv1992** [comment]
Ok, I don't see most of these as improvements to be honest, just different and in most cases slightly less readable.

File: src/networkaddress.rs [hunk 1]
```diff
@@ -59,17 +59,22 @@ impl NetworkAddress for SocketAddrV4 {
...
-        out.sin_addr = libc::in_addr {
-            s_addr: u32::from_ne_bytes(self.ip().octets()),
+        let sockaddr = libc::sockaddr_in {
+            sin_family: libc::AF_INET as _,
+            sin_port: u16::from_ne_bytes(self.port().to_be_bytes()),
+            sin_addr: libc::in_addr {
+                s_addr: u32::from_ne_bytes(self.ip().octets()),
+            },
+            sin_zero: [0; 8],
         };
```
Comment (folkertdev) [inline src/networkaddress.rs:69]: show clearly that the whole struct is initialized

File: src/networkaddress.rs [hunk 2]
```diff
@@ -59,17 +59,22 @@ impl NetworkAddress for SocketAddrV4 {
...
 
+        // # Safety
+        //
+        // The assertions above guarantee the validity of the cast. The asserts and the fact that
+        // result is allocated on the stack guarantee the validity of the write
+        let mut result = zeroed_sockaddr_storage();
+        unsafe { std::ptr::write(&mut result as *mut _ as *mut libc::sockaddr_in, sockaddr) };
```
Comment (folkertdev) [inline src/networkaddress.rs:76]: don't even bother with making a reference and the potential safety implications that has; just write to a pointer

File: src/networkaddress.rs [hunk 3]
```diff
@@ -107,28 +126,27 @@ impl MulticastJoinable for SocketAddrV4 {
...
             },
             imr_address: libc::in_addr {
                 s_addr: u32::from_ne_bytes(Ipv4Addr::UNSPECIFIED.octets()),
             },
-            imr_ifindex: InterfaceDescriptor {
-                interface_name: Some(interface),
-                mode: crate::interface::LinuxNetworkMode::Ipv4,
-            }
-            .get_index()
-            .ok_or(std::io::ErrorKind::InvalidInput)? as _,
```
Comment (folkertdev) [inline src/networkaddress.rs:122]: this formats in a (to me) confusing way where it looks like you assign a `InterfaceDescriptor` literal to the `imr_ifindex` field. But no, there are some sneaky method calls that change the whole meaning. This should just be restructured in any case. (I know we want to change this anyway, but still)

File: src/networkaddress.rs [hunk 4]
```diff
@@ -107,28 +126,27 @@ impl MulticastJoinable for SocketAddrV4 {
...
+        // # Safety
+        //
         // value points to a struct of length option_len, of type ip_mreq as expected for IPPROTO_IP/IP_ADD_MEMBERSHIP
         cerr(unsafe {
             libc::setsockopt(
                 socket,
                 libc::IPPROTO_IP,
                 libc::IP_ADD_MEMBERSHIP,
-                &request as *const _ as *const _,
+                &request as *const libc::ip_mreqn as *const libc::c_void,
```
Comment (folkertdev) [inline src/networkaddress.rs:149]: explicit is better than implicit. Otherwise you're just guessing types, and copy/pasting might invalidate the assumptions made by the `_`

File: src/socket.rs [hunk 5]
```diff
@@ -156,7 +156,7 @@ impl<A: NetworkAddress, S> Socket<A, S> {
     ) -> std::io::Result<Option<Timestamp>> {
         const TIMEOUT: Duration = Duration::from_millis(200);
 
-        if let Ok(timestamp) = tokio::time::timeout(TIMEOUT, async {
+        let fut = async {
```
Comment (folkertdev) [inline src/socket.rs:159]: a 10+ line expression in the if condition... just looks weird and is hard to parse

Comment (davidv1992) [reply src/socket.rs:159]: For me, the if expression more clearly shows what the code is doing, instead of having the code and its execution completely separated. I am not saying there ain't ways to improve this, there may well be, but for me this isn't it.

File: src/networkaddress.rs (hunk 4 above)
Comment (davidv1992) [reply src/networkaddress.rs:149]: No, the explicit here is just distraction and repeating type names don't magically make them more correct. What the code needs to do is to provide a reference to the request, just encoded as a slightly inconvenient type which the first line gets us to. The reference to request part is what I think the core of this line is, not the unfortunate conversion that is needed.

File: src/networkaddress.rs (hunk 3 above)
Comment (davidv1992) [reply src/networkaddress.rs:122]: This is indeed somewhat convoluted, but actually the interfacedescriptor assignment is really close to the intended meaning here, namely that we want it to use the provided interface. The function calls really are to deal with the fact that a conversion is needed.

File: src/networkaddress.rs (hunk 1 above)
Comment (davidv1992) [reply src/networkaddress.rs:69]: Ok, for me this feels weird because what we want is not a sockaddr_in. We want a sockaddr(_storage) containing an ipv4 address, which for me the original formulation captures better.

File: src/networkaddress.rs (hunk 2 above)
Comment (davidv1992) [reply src/networkaddress.rs:76]: see above
